tensorflow/compiler/xla/service/depthwise_convolution_converter.cc

```diff
@@ -0,0 +1,212 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/depthwise_convolution_converter.h"
+
+#include <memory>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "tensorflow/compiler/xla/literal.h"
+#include "tensorflow/compiler/xla/literal_util.h"
+#include "tensorflow/compiler/xla/service/dfs_hlo_visitor_with_default.h"
+#include "tensorflow/compiler/xla/service/hlo_computation.h"
+#include "tensorflow/compiler/xla/service/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/hlo_opcode.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+#include "tensorflow/compiler/xla/types.h"
+#include "tensorflow/compiler/xla/util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace xla {
+
+namespace {
+
+class ConvolutionVisitor : public DfsHloVisitorWithDefault {
+ public:
+  // Default visitor action is to do nothing and return OK.
+  Status DefaultAction(HloInstruction* /*hlo_instruction*/) override {
+    return Status::OK();
+  }
+
+  Status HandleConvolution(HloInstruction* convolution) override;
+
+  Status HandleBackwardFilterBatchGroupConvolution(HloInstruction* convolution);
+
+  // Runs the visitor on a computation.
+  static bool Run(HloComputation* computation,
+                  std::function<bool(HloInstruction*)> is_cost_viable);
+
+  // Returns whether any convolution ops were rewritten.
+  const bool changed() const { return changed_; }
+
+  ~ConvolutionVisitor() override = default;
+
+ private:
+  explicit ConvolutionVisitor(
+      HloComputation* computation,
+      std::function<bool(HloInstruction*)> is_cost_viable)
+      : computation_(computation), is_cost_viable_(is_cost_viable) {}
+
+  // Current HloComputation instance the ConvolutionVisitor is traversing.
+  HloComputation* computation_;
+
+  // Whether rewrite has occurred.
+  bool changed_ = false;
+
+  std::function<bool(HloInstruction*)> is_cost_viable_;
+};
+
+bool ConvolutionVisitor::Run(
+    HloComputation* computation,
+    std::function<bool(HloInstruction*)> is_cost_viable) {
+  ConvolutionVisitor visitor(computation, is_cost_viable);
+  TF_CHECK_OK(computation->Accept(&visitor));
+  return visitor.changed_;
+}
+
+Shape SwapInputOutputFeatureDims(const Shape& shape, int64 input_feature_dim,
```

Comment:
Please put this function into an anonymous namespace

namespace {
Shape SwapInputOutputFeatureDims(...
}  // namespace